pkg/controllers/disruption/consolidation.go

```diff
@@ -177,10 +181,7 @@ func (c *consolidation) computeConsolidation(ctx context.Context, candidates ...
 
 	if allExistingAreSpot &&
 		results.NewNodeClaims[0].Requirements.Get(v1beta1.CapacityTypeLabelKey).Has(v1beta1.CapacityTypeSpot) {
-		if len(candidates) == 1 {
-			c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "Can't replace a spot node with a spot node")...)
-		}
-		return Command{}, nil
+		return c.computeSpotToSpotConsolidation(ctx, candidates, results, instanceTypesOptionsFilteredByPrice)
```

Comment:
An edge case I just realized that we have here: The way we've written this, we won't try a consolidation to an on-demand instance type if we can attempt a spot-to-spot consolidation. That spot-to-spot consolidation may always fail, but we might be able to get an on-demand instance type (where we are more flexible to less options). It's less likely that we will consolidate from spot to on-demand, but we should be correct here to ensure that we can still consolidate from spot to on-demand if we are flexible to both and the spot-to-spot consolidation isn't possible but the spot-to-on-demand one is

Reply:
I am little confused as to whether we have the spot to on-demand consolidation today ? I thought we only had "delete" consolidation for spot right ? 

Are we fearing the possibility of every possible spot replacement instance being ICE'd and we only have on-demand instances left that are cheaper than the current spot candidate and we have to replace the spot candidate with on-demand instance ?

Reply:
We have the ability to spot to on-demand if there are multiple node pools, just not within one node pool.

Reply:
I am not sure how is that possible today when we always return empty command for spot candidate [here](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/controllers/disruption/consolidation.go#L178-L183)

I think the only possibility for this to happen would be to change the requirement of NodePool from only spot to only od in which case, the check above would fail for the hasSpot in requirements and then would be replaced by the od node. Correct me if I am wrong.

Reply:
I think we discussed this, and this should be resolved, right @nikmohan123? 

The code you linked is for if all nodes are spot nodes, and the scheduling simulation's node claim has spot. If we have another nodepool that has only on-demand, this would consolidate.

Reply:
Yes, we discussed this and I think we have not changed the behavior with respect to this. We can resolve this.